core/ipc/grpc/client/src/main/java/org/opennms/core/ipc/grpc/client/MinionGrpcClient.java

```diff
@@ -0,0 +1,509 @@
+                }
+            }
+        } else {
+            LOG.info("gRPC IPC server is not in ready state");
+        }
+        return false;
+    }
+
+
+    private void sendMinionHeaders() {
+        RpcResponseProto rpcHeader = RpcResponseProto.newBuilder()
+                .setLocation(minionIdentity.getLocation())
+                .setSystemId(minionIdentity.getId())
+                .setModuleId(MINION_HEADERS_MODULE)
+                .setRpcId(minionIdentity.getId())
+                .build();
+        sendRpcResponse(rpcHeader);
+        LOG.info("Sending Minion Headers from SystemId {} to gRPC server", minionIdentity.getId());
+    }
+
+    private void processRpcRequest(RpcRequestProto requestProto) {
+        long currentTime = requestProto.getExpirationTime();
+        if (requestProto.getExpirationTime() < currentTime) {
+            return;
+        }
+        String moduleId = requestProto.getModuleId();
+        if (Strings.isNullOrEmpty(moduleId)) {
+            return;
+        }
+        LOG.debug("Received RPC request with RpcID:{} for module {}", requestProto.getRpcId(), requestProto.getModuleId());
+        RpcModule<RpcRequest, RpcResponse> rpcModule = registerdModules.get(moduleId);
+        if (rpcModule == null) {
+            return;
+        }
+        //Build child span from rpcMessage and start minion span.
+        Tracer.SpanBuilder spanBuilder = buildSpanFromRpcMessage(requestProto);
+        Span minionSpan = spanBuilder.start();
+        setTagsForRpc(requestProto, minionSpan);
+
+        RpcRequest rpcRequest = rpcModule.unmarshalRequest(requestProto.getRpcContent().toStringUtf8());
+        CompletableFuture<RpcResponse> future = rpcModule.execute(rpcRequest);
+        future.whenComplete((res, ex) -> {
+            final RpcResponse rpcResponse;
+            if (ex != null) {
+                // An exception occurred, store the exception in a new response
+                LOG.warn("An error occured while executing a call in {}.", rpcModule.getId(), ex);
+                rpcResponse = rpcModule.createResponseWithException(ex);
+                minionSpan.log(ex.getMessage());
+                minionSpan.setTag(TAG_RPC_FAILED, "true");
+            } else {
+                // No exception occurred, use the given response
+                rpcResponse = res;
+            }
+            minionSpan.finish();
+            // Construct response using the same rpcId;
+            String responseAsString = rpcModule.marshalResponse(rpcResponse);
+            RpcResponseProto responseProto = RpcResponseProto.newBuilder()
+                    .setRpcId(requestProto.getRpcId())
+                    .setSystemId(minionIdentity.getId())
+                    .setLocation(requestProto.getLocation())
+                    .setModuleId(requestProto.getModuleId())
+                    .setRpcContent(ByteString.copyFrom(responseAsString.getBytes()))
```

Comment:
```suggestion
                    .setRpcContent(ByteString.copyFrom(responseAsString, StandardCharsets.UTF_8))
```

avoid copying bytes twice.